torch/utils/checkpoint.py

```diff
@@ -275,3 +275,81 @@ def forward(input):
         input = checkpoint(run_function(start, end, functions), input,
                            preserve_rng_state=preserve)
     return run_function(end + 1, len(functions) - 1, functions)(input)
+
+
+class Checkpoint():
+    """Checkpoitining without re-entrant autograd
+
+    Args:
+        function: describes what to run in the forward pass of the model or
+            part of the model. It should also know how to handle the inputs
+            passed as the tuple. For example, in LSTM, if user passes
+            ``(activation, hidden)``, :attr:`function` should correctly use the
+            first input as ``activation`` and the second input as ``hidden``
+        preserve_rng_state(bool, optional, default=True):  Omit stashing and restoring
+            the RNG state during each checkpoint.
+    """
+    def __init__(self, function, preserve_rng_state=True):
+        self.function = function
+        self.preserve_rng_state = preserve_rng_state
+
+    def __call__(self, *args):
+        check_backward_validity(*args)
+        self.had_autocast_in_fwd = torch.is_autocast_enabled()
+
+        if self.preserve_rng_state:
+            self.fwd_cpu_state = torch.get_rng_state()
+            # Don't eagerly initialize the cuda context by accident.
+            # (If the user intends that the context is initialized later, within their
+            # run_function, we SHOULD actually stash the cuda state here.  Unfortunately,
+            # we have no way to anticipate this will happen before we run the function.)
```

Comment:
I think we can actually detect that after the user function ran and raise a nice error about it